csrc/scheduler/hopper_multi_matmul.cpp

```diff
@@ -421,106 +421,52 @@ void HopperMultipleMatmulScheduler::scheduleMmaResults() {
-      gemm_tile.cta_tile.m,
-      ", actual: ",
-      tile_size_m);
-  NVF_ERROR(
-      tile_size_n == gemm_tile.cta_tile.n,
-      "Actual tile size at axis(-1) in output tensor is different from CTA tile size! Expected: ",
-      gemm_tile.cta_tile.n,
-      ", actual: ",
-      tile_size_n);
-  const int64_t tot_elements = tile_size_m * tile_size_n;
-  constexpr int64_t warp_size = 32l;
-  const int64_t tidx = warp_size;
-  const int64_t tidy = gemm_tile.cta_tile.n / gemm_tile.warp_tile.n;
-  const int64_t tidz = gemm_tile.cta_tile.m / gemm_tile.warp_tile.m;
-  // step-1, merge last 2 dims
-  c->merge(-2);
-  // [Mo, No, m*n]
-
-  // step-2, set vectorization to maximum
-  // We have fixed tidx, tidy, and tidz, so we need to make sure that the
-  // output tensor is divisible by tidx * tidy * tidz * vectorization_factor
-  NVF_ERROR(
-      tot_elements % (tidx * tidy * tidz * vectorization_factor) == 0,
-      "Output tensor cannot be fully vectorized! tot_elements:",
-      tot_elements,
-      ", tidx: ",
-      tidx,
-      ", tidy: ",
-      tidy,
-      ", tidz: ",
-      tidz,
-      ", vectorization_factor: ",
-      vectorization_factor);
-  c->split(-1, vectorization_factor);
-  c->axis(-1)->parallelize(ParallelType::Vectorize);
-  // [Mo, No, m*n/vect, vect]
-
-  // step-3, Split out a warp for TIDx
-  c->split(-2, tidx);
-  c->axis(-2)->parallelize(ParallelType::TIDx);
-  // [Mo, No, m*n/vect/TIDx, TIDx, vect]
-
-  // step-4, Split out for TIDy and TIDz
-  // TIDy = cta_tile_n/warp_tile_n
-  // TIDz = cta_tile_m/warp_tile_m
-  c->split(-3, tidy);
-  c->axis(-3)->parallelize(ParallelType::TIDy);
-
-  c->split(-4, tidz);
-  c->axis(-4)->parallelize(ParallelType::TIDz);
-  // [Mo, No, m*n/vect/TIDx/TIDy/TIDz, TIDz, TIDy, TIDx, vect]
-
-  for (TensorView* mma_result : mma_results_) {
-    // step-5, Parallel first 2 dims same as mma_result
-    scheduler_utils::parallelizeAllLike(
-        mma_result,
-        2,
-        {c},
-        {ParallelType::BIDx, ParallelType::BIDy, ParallelType::BIDz});
-  }
+  // Block Schedule and Parallelize
+  blockTileTensors({c});
+  parallelizeBlocks({c});
+
+  // Apply mma common transformation
+  c->split(-2, getM(params_->mma_macro));
+  c->split(-1, getN(params_->mma_macro));
+  // [..., Mo, No, Mio, Mii, Nio, Nii]
+  // -> [..., Mo, No, Mio, Nio, Mii, Nii]
+  c->reorder({{-3, -2}});
+  c->merge(-4);
+
+  // [...., Mii, Nii] ->
+  // [...,  Mii/16,  Miioi(2), Miii(8), Nii/8, Niio(4), Niii(2)] ->
+  // [.., Mii/16,  Miii(8), Niio(4),  Nii/8, Miioi(2), Niii(2)   ]
+  // [..., Mii/16 * Miii(8) * Niio(4), Nii/8, Miioi(2), Niii(2)  ]
+  // Mii/16 * Miii(8) * Niio(4) is 128 for Hopper and this is parallelized as
+  // TIDx.
+  auto s =
+      mma_utils::MmaSwizzler::scheduleMmaOutputAllocation(c->getLoopDomain());
+  c->setLoopDomain(s.as<IterDomain*>());
+  c->axis(-5)->parallelize(ParallelType::TIDy);
```

Comment:
I mentioned in https://github.com/NVIDIA/Fuser/pull/3575#discussion_r1882043814 that this whole chunk should basically be a new method.

Reply:
I'll wait for @rdspring1 to land his PR then use his fix or add one.